[ Hardik Narang ] iP

naranghardik16 opened the pull request:
# Duke

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://medium.com/the-mission/the-most-popular-productivity-pieces-of-wisdom-from-david-allen-72ffe70ac7b9)) 

Duke frees your mind of having to remember things you need to do. It's
* text-based
* ~~EASY~~ SUPER EASY to learn
* SUPER FAST to use

All you need to do is, 
1. add your tasks
2. let it manage your tasks for you :wink:
3. download it from [here](https://github.com/naranghardik16/ip).

And it is **FREE!**

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too.  Here's the `main` method:

```java
public static void main(String[] args) {
      new Duke(DUKE_TXT).runDuke();
}
```

Commit:
Add Gradle support

naranghardik16 committed:
First commit (Level-1)

naranghardik16 committed:
Added Level-2 increment

naranghardik16 committed:
Added Level-3 increment

naranghardik16 committed:
Added Level-4 Increment

naranghardik16 committed:
Added Exception handling

naranghardik16 committed:
Added Level-6 increment

naranghardik16 committed:
automate Ui testing

naranghardik16 committed:
Made minor changes (worked on organization)

naranghardik16 committed:
Level-7

naranghardik16 committed:
Made minor changes

naranghardik16 committed:
Revert "Level-7"

This reverts commit 7ea4cc34e50773222962d8eaf80a14ce04e95bbc.

naranghardik16 committed:
Made minor changes

naranghardik16 committed:
Level-8

naranghardik16 committed:
Resolving conflicts

naranghardik16 committed:
Resolving conflict

naranghardik16 committed:
Merge branch 'branch-Level-7'

naranghardik16 committed:
Merge branch 'branch-Level-8'

LZher385 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,275 @@
+    public static void readInput(String userInput){
+
+        // Get command from the userInput
+        String[] splitLine = userInput.split(" ", 2);
```
Shouldn't a plural noun be used for the list of substrings?

LZher385 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,16 @@
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task{
+    private LocalDateTime at;
```
I know its easy to tell from the context, but maybe the variable name can be a little more descriptive? I noticed this in `Deadline.java` too 👍

nobelang78 commented on data/duke.txt:
```diff
@@ -0,0 +1,8 @@
+T | 1 | new
```
Should the save files be uploaded to github?

NayLin-H99 replied:
You should leave duke.txt in the .gitignore file.

nobelang78 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,275 @@
+            throw new DoneAlreadyException();
+        }
+
+        // Mark the index as done
```
Perhaps proper javadocs could be added so IDE can display the documentation with all the information?

nobelang78 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,275 @@
+        userInput = userInput.replaceAll( RegexType.START_LINE_REGEX.getRegexType() + command + RegexType.SPACE_REGEX.getRegexType(), "");
+
+        // Check the command type then execute the commands
+        if(userInput.equals(InputType.LIST.getInputType())) {
```
Could this list of checks better be served with a switch statement to improve readability?

codecodecoder78 reviewed (commented):
Generally great job for abstraction, few nits on styling to fix for clarity!

naranghardik16 committed:
A-MoreOOP

naranghardik16 committed:
Made minor changes

naranghardik16 committed:
Merge branch 'add-gradle-support'

naranghardik16 committed:
Added support for testing

naranghardik16 committed:
Added JavaDoc

naranghardik16 committed:
Added JavaDoc

naranghardik16 committed:
Added CodingStandard

naranghardik16 committed:
Added Level-9

naranghardik16 committed:
Merge branch 'branch-A-CodingStandard'

naranghardik16 committed:
Merge branch 'branch-Level-9'

naranghardik16 committed:
Merged all 3 branches and made minor changes

NayLin-H99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,79 @@
+import duke.InputType;
+import duke.Parser;
+import duke.Storage;
+import duke.tasks.*;
```
Wildcard Imports are suggested against in this module. Perhaps you can set your IDE to not "optimise" imports automatically.
![image](https://user-images.githubusercontent.com/77244987/131891531-f5237542-92cd-45e9-be4a-6410c4cb321f.png)

NayLin-H99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,79 @@
+        try {
+            return command.execute();
+        } catch (Exception e) {
+            return new CommandResult(e.getMessage(), false, null);
```
Did you not overload your constructor? Do you need to input the "null" argument here?

NayLin-H99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,79 @@
+        userInput = in.nextLine();
+
+        // While the input is not "bye", add input to array of tasks and get another input
+        while (!userInput.equals(InputType.BYE.getInputType())) {
```
I think the readibility of this line can be improved by using the "bye" string directly or abstracting this into an ExitCommand as suggested under the specifications. Don't you think so?

NayLin-H99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+        final String arguments = userInput.replaceAll( RegexType.START_LINE_REGEX.getRegexType() + command + RegexType.SPACE_REGEX.getRegexType(), "");
+
+        // Check the command type then execute the commands
+        if(arguments.equals(InputType.LIST.getInputType())) {
+            return new ListCommand(taskList);
+        }
```
Doing this kind of defeats the purpose of using enum, right? Instead of using enum purely as strings, I suggest parsing the user input into enum, then using the enum to return the appropriate commands. 

An additional suggestion is to use switch case, which is recommended for enum since you can do something like: 
```suggestion
        switch (inputType) {
        case LIST: 
            return new ListCommand(taskList);
```

NayLin-H99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+        }
+        return new EmptyCommand(taskList);
+    }
+}
```
At the end of every file, you should leave a newline! 😄

NayLin-H99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,142 @@
+package duke;
+
+import duke.tasks.*;
```
Once again, try to adjust your IDE to not optimise your imports 😄

NayLin-H99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,142 @@
+                case "E":
+                    taskList.tasks.add(new Event(description, atBy, isDone));
+
+                    break;
```
Indentation is not according to the standards. Perhaps you can adjust your IDE settings for this too?
![image](https://user-images.githubusercontent.com/77244987/131895324-61941270-a7a7-4607-ac33-577eedc84167.png)

NayLin-H99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,142 @@
+        TaskManager taskList = new TaskManager();
+        String message = "";
+
+        try {
+            loadData(taskList);
+        } catch (FileNotFoundException e){
+            message = message + "File not found";
+            try {
+                createFile();
+                message = message + "\n" + "New file data/duke.txt created";
+
+            } catch (IOException e1){
+                message = message + "\n" + "Error creating new file";
+            }
+
+        } catch (FileFormatException e){
+            message = message + "\n" + "File formatting error";
+            taskList.clearList();
+        }
```
I strongly suggest against nesting try catch blocks. You should avoid write business logic inside error handling. Perhaps you can let `loadData()` return some boolean and then check whether to `createFile()` from there?

NayLin-H99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,37 @@
+                + "---------------------------------------");
+    }
+
+    public void printOneLine(String line){
```
You should `private` this method since you are abstracting this to other print methods, right?

NayLin-H99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,79 @@
+                String errorMessage = storage.writeToFile(taskList);
+
+                ui.printMessage(errorMessage);
+            }
```
I don't think you need to check for this. You can save the file after every command.

NayLin-H99 commented on src/main/java/duke/command/EmptyCommand.java:
```diff
@@ -0,0 +1,16 @@
+import duke.PrintType;
+import duke.tasks.TaskManager;
+
+public class EmptyCommand extends Command{
```
I don't think you need to create a class just to handle an empty command? I suggest just throwing a DukeException for unrecognised strings, since you are using Regex anyways.

NayLin-H99 commented on src/main/java/duke/tasks/TaskManager.java:
```diff
@@ -0,0 +1,223 @@
+     * @return - the message of acknowledgement in String form
+     * @throws TodoException - if the description is empty
+     */
+    public String addTodo(String userInput) throws TodoException {
```
I'm not sure what is the point of having individual command classes where you are doing all the commands under the Task Manager itself?

NayLin-H99 commented on src/main/java/duke/tasks/TaskManagerUi.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Class to maintain the UI of the list of tasks
+ */
+public class TaskManagerUi {
```
Your UI class should handle interactions with the user? Is this not just a tuple?

NayLin-H99 commented on src/main/java/duke/tasks/Todo.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public class Todo extends Task {
+    private TaskType type = TaskType.TODO;
+    public Todo (String description){
```
Try to have consistent formatting 😄 
```suggestion
    public Todo (String description) {
```

NayLin-H99 reviewed (commented):
I really liked that you have done a lot of abstraction in your code.

naranghardik16 committed:
Added GUI capabilities

naranghardik16 committed:
Merge branch 'branch-Level-10'

naranghardik16 committed:
Improve Code Quality

naranghardik16 committed:
Improve Code Quality

Ensure the method/class header comments follow the format specified in the coding standard.

Points improved:
- The Javadoc comment is minimal.
- Include tags wherever required and follow the particular order.

naranghardik16 committed:
Improve formatting

naranghardik16 committed:
Add Assertions

Make use of assert feature to document important assumptions that should hold at various points in the code.

Refer to this for more details on Java assertions:
https://www.geeksforgeeks.org/assertions-in-java/

naranghardik16 committed:
Merge pull request #3 from naranghardik16/branch-A-CodeQuality

Merging branch-A-CodeQuality

naranghardik16 committed:
Merge branch 'branch-A-Assertions'

naranghardik16 committed:
Merge pull request #2 from naranghardik16/branch-A-Assertions

Add Assertions

naranghardik16 committed:
Add Extension C-Priority

Provide a way to attach priorities to tasks such as Low or High
This will help the user to prioritize the tasks and become more efficient
A task with High priority needs to be done first
A task with Low priority needs to be done at the last.

Have implemented the extension C-Priority

Refer to the other list of extensions here:
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/projectDuke/index.html#extensions-category-b

naranghardik16 committed:
Merge branch 'branch-C-Priority'

naranghardik16 committed:
Merge pull request #4 from naranghardik16/branch-C-Priority

Branch c priority

naranghardik16 committed:
Merge branch 'master' of https://github.com/naranghardik16/ip

naranghardik16 committed:
Make changes to text file

naranghardik16 committed:
Add representative screenshot

naranghardik16 committed:
Add User Guide

naranghardik16 committed:
Set theme jekyll-theme-minimal

naranghardik16 committed:
Set theme jekyll-theme-cayman

naranghardik16 committed:
Add UserGuide

naranghardik16 committed:
Set theme jekyll-theme-cayman
